### ISSUE TITLE
[Sync Enhancement | Data Access Checker] Data filtering based user Sync Locations and Client Location

### ISSUE THREAD

**Issue by dubdabasoduba** [opensrp/fhircore#3052]
**Keyword Definition**
- **Client** - An entity that has data collected against it (e.g. Family/Household, Family/Household member, Patient, Service Point or Product).
- **Client Location** - This term is used to refer to location(s) associated with the client.
    - This would mean the location where the family or family member lives for the case of the SID (Summit Institute for Development) applications
    - This would mean the service point of where data collected belongs to for the case of EUSM (End User Supply Monitoring)
- **Sync Locations** - These are the locations used to filter the data to be synced to the application.

**Context**
- Sync by Client Location means getting all the resources that have a Client Location ID that is a child or equal to the user sync locations. 
- The generation of the tags to filter by would work almost similar to the Sync by Practitioner Location https://github.com/opensrp/fhircore/issues/1662
- The Sync Locations will be part of the request parameters. These will be found on the `syncLocation` param. 
- We can find the child locations by using the Location Hierarchy functionality.  
- If the Client Location `userAssignedLocation`deployment config is set to `true` 
    - Fetch the User Assigned Location and user it as the sync location   

**Technical Description of enhancement**
- [ ] From the request parameters get the Sync locations from the `syncLocation` param.
- [ ] For each Sync Location fetch its hierarchy and fetch the Location IDs of all its children. 
    - [ ] Add the Children Location IDs + the Sync Location IDs to an array  
- [ ] Use the array created above to create query parameters for the FHIR Store e.g. HAPI then use these queries to fetch and filter data from HAPI
    - https://<base_url>/fhir/Patient?_tag=137151 -- Location 1
    - https://<base_url>/fhir/Patient?_tag=136256 --Location 2
- [ ] Remember to add the other pagination request params to each request. 
- [ ] Bundle the responses from each of the queries into one response and send that back. 
   - [ ] This Bundle should match the FHIR resource Bundle specifications. 

**Acceptance Criteria**
- [ ] This functionality should only kick in if the sync strategy **Client Location** is used 
- [ ] Return only the resources that have a Client Location ID `meta.tag` that is a child or equal to the user sync locations. 
- [ ] If the Client Location `userAssignedLocation`deployment config is set to `true`, fetch the User Assigned Location and user it as the sync location

**More information on why we need to make this change**
- https://docs.google.com/document/d/1D7fXYo3zFMAO3j5Vh8FmpfSVeRfxfVjHAefDtw2KmSk/edit?usp=sharing

**Comment by pld** [comment]
I don't think the client should be responsible for constructing the location list, I think the gateway should, that way we reduce the round trips and we have that business logic on the server side, so we can better control it. I.e. the flow is

- Fetch sync locations from the `syncLocations` param
- Send sync request to a new gateway endpoint with
    - syncLocations array
    - metaTag to sync on (e.g. practioner-location-at-write-time entity-location-at-write-time)
    - pagination

The logic in the gateway endpoint converts the locations array into a set of locations

**Comment by dubdabasoduba** [comment]
> I don't think the client should be responsible for constructing the location list, I think the gateway should, that way we reduce the round trips and we have that business logic on the server side, so we can better control it. I.e. the flow is
> 
> * Fetch sync locations from the `syncLocations` param
> * Send sync request to a new gateway endpoint with
>   
>   * syncLocations array
>   * metaTag to sync on (e.g. practioner-location-at-write-time entity-location-at-write-time)
>   * pagination
> 
> The logic in the gateway endpoint converts the locations array into a set of locations

@pld  that's the idea

The client will only be responsible for sending the `syncLocations` param. The `syncLocation` are the locations selected by the user on the app, for microplanning apps or the logged in user assigned location for health workflow apps

The gateway will then be responsible for fetching the other child locations internally

**Comment by pld** [comment]
OK, that's not clear to me from the technical description in this issue, this reads like the client is doing that, can we move that part of this issue out to an issue in the info gateway repo

**Comment by ndegwamartin** [comment]
Looks good, a few thoughts
- The entry in the json app configs for the new sync strategy should be the pascal cased literal `'RelatedEntityLocation'` to match the others.
- On the role we can invert the concept so that it reflect exactly the grants e.g. `ROLE_ALL_LOCATIONS` or `ROLE_ALL_LOCATION_DATA`, `ROLE_REL_ALL_LOCATIONS` or `ROLE_RELATED_ENTITY_LOCATIONS_ALL`
- When constructing the location ids data structure when processing the hierarchy we should use a `SET` to avoid duplicates

**Comment by pld** [comment]
Cool I like the exact grant idea, it's unclear what the options are, it seems like they're just
- `ROLE_ALL_LOCATIONS`
- `ROLE_ASSIGNED_LOCATIONS`

b/c we're not saying anything specific about the sync strategy in the grant, only whether the user has access to all or only their assigned locations and sub-nodes.

**Comment by dubdabasoduba** [comment]
@pld @ndegwamartin we can use `ROLE_ASSIGNED_LOCATIONS`

**Comment by ndegwamartin** [comment]
@pld @dubdabasoduba yeah since the two will be mutually exclusive, we should only pick one of the roles . 

For the two options above, I'd go with `ROLE_ALL_LOCATIONS` because IMO it authorises ability to view all locations _(in line with my understanding of roles)_ rather than `ROLE_ASSIGNED_LOCATIONS` which acts more as a flag

**Comment by pld** [comment]
I'm good with that, makes senseOn Feb 29, 2024, at 04:40, Martin Ndegwa ***@***.***> wrote:﻿
@pld @dubdabasoduba yeah since the two will be mutually exclusive, we should only pick one of the roles .
For the two options above, I'd go with ROLE_ALL_LOCATIONS because IMO it authorises ability to view all locations (in line with my understanding of roles) rather than ROLE_ASSIGNED_LOCATIONS which acts more as a flag
